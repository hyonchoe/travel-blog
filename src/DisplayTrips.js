import React, { useState, useEffect, useRef } from 'react'
import { Empty, Skeleton, Row, Col, BackTop, message, Modal, Button, List } from 'antd'
import { useAuth0 } from '@auth0/auth0-react'

import tripService from './services/tripService.js'
import MyMapContainer from './MyMapContainer.js'
import Trip from './Trip'

const DisplayTrips = (props) => {
    const [tripList, setTripList] = useState({
      trips: [],
      loadingData: true,
      noMoreRecords: false,
    })
    
    const [modalMap, setModalMap] = useState({
      modalVisible: false,
      tripLocations: null,
    })

    const [scrollToTripId, setScrollToTripId] = useState('')
    const lastItemRef = useRef(null)

    const { getAccessTokenSilently, isAuthenticated, user } = useAuth0()
    const displayLimit = 100

    useEffect(() => {
      const fetchData = async () => {
        const res = (props.showMyTrips) ?
          await tripService.getTrips(getAccessTokenSilently)
          : await tripService.getPublicTrips(null)

        setTripList( {
          ...tripList,
          trips: res,
          loadingData: false,
        })
      }

      fetchData()
    }, [])

    const reloadTripData = async () => {
      setTripList( {
        ...tripList,
        loadingData: true,
      })

      const res = await tripService.getPublicTrips(null)

      setTripList( {
        trips: res,
        loadingData: false,
        noMoreRecords: false,
      })
    }

    const onLoadMore = async () => {
      setTripList({
        ...tripList,
        loadingData: true,
      })
      
      const res = await tripService.getPublicTrips(tripList.trips[tripList.trips.length-1])

      let updatedData = tripList.trips.concat(res)
      let updateScrollPos = false
      if (updatedData.length > displayLimit){
        setScrollToTripId(tripList.trips[tripList.trips.length-1]._id)
        updateScrollPos = true
        
        updatedData = updatedData.slice(updatedData.length - displayLimit)
        updatedData[0] = { placeholder: true }
      }
      const noMoreRecords = res.length === 0 || res[res.length-1].noMoreRecords
      
      setTripList({
        trips: updatedData,
        loadingData: false,
        noMoreRecords: noMoreRecords,
      })
      if (noMoreRecords){
        message.info('No more older public trips to display')
      }

      if (updateScrollPos && lastItemRef.current){
        lastItemRef.current.scrollIntoView()
      }
    }

    const handleDeleteTrip = async (tripId) => {
      const res = await tripService.deleteTrip(tripId, getAccessTokenSilently)
      console.log(res)
      
      let tripTitle = ''

      setTripList({
        ...tripList,
        trips: tripList.trips.filter((trip) => {
                  if (trip._id === tripId){
                      tripTitle = trip.title
                  }

                  return trip._id !== tripId
              })
      })

      message.success(`Trip "${tripTitle}" removed successfully`)
    }
    const handleLaunchMapModal = (tripTitle, tripLocations) => {
      setModalMap({
        modalVisible: true,
        tripLocations: tripLocations
      })
    }
    const handleModalOk = () => {
      setModalMap({
        modalVisible: false,
        tripLocations: null
      })
    }

    const tripCountsPerSize = 10
    const showMyTrips = props.showMyTrips
    const handleEditTrip = props.editTrip
    let mapCenterLat = null
    let mapCenterLng = null
    if (modalMap.tripLocations && modalMap.tripLocations.length>0) {
      mapCenterLat = modalMap.tripLocations[0].latLng[0]
      mapCenterLng = modalMap.tripLocations[0].latLng[1]
    }
    const userId = (isAuthenticated) ? user.sub : ''

    const loadMoreButton = (!tripList.loadingData && !tripList.noMoreRecords) ? 
                              (<div style={{ textAlign: 'center', marginTop: 12, }} >
                                <Button onClick={onLoadMore}>Load more</Button>
                              </div>)
                              : null
    const tripCard = (item, itemRef) =>
                              (<div ref={itemRef}>
                                { item.placeholder && 
                                <div style={{ textAlign: 'center', }} >
                                  <Button onClick={reloadTripData}>Trips removed from display. Click to refresh the feed</Button>
                                </div>
                                }
                                { !item.placeholder &&
                                <Trip
                                  isAuthenticated={isAuthenticated}
                                  userId={userId}
                                  showMyTrips={showMyTrips}
                                  trip={item}
                                  deleteTrip={handleDeleteTrip}
                                  editTrip={handleEditTrip}
                                  launchMapModal={handleLaunchMapModal} />
                                }
                              </div>)

    return (
      <div className="displayTripsContainer">
      <BackTop />
      <Row
        gutter={[8, 16]}
        justify="center" >
        <Col span={4} />
        <Col span={16}>
          { !showMyTrips && tripList.trips.length > 0 && 
          <List
            itemLayout="vertical"
            loading={tripList.loadingData}
            loadMore={loadMoreButton}
            dataSource={tripList.trips}
            renderItem={(item) => (
              <List.Item>
                {item._id === scrollToTripId && 
                  tripCard(item, lastItemRef)
                }
                {item._id !== scrollToTripId && 
                  tripCard(item, null)
                }
              </List.Item>
            )} />
          }
          { showMyTrips && tripList.trips.length > 0 && 
          <List
            itemLayout="vertical"
            pagination={{
              onChange: page=> {
                window.scrollTo({top:0, behavior: 'smooth'})
              },
              pageSize: tripCountsPerSize,
            }}
            dataSource={tripList.trips}
            renderItem={(item) => (
              <List.Item>
                {tripCard(item)}
              </List.Item>
            )} />
          }
          { tripList.loadingData &&
          <Skeleton loading={true} active />
          }
          { tripList.trips.length === 0 && !tripList.loadingData &&
          <Empty />
          }
        </Col>
        <Col span={4} />
      </Row>
      <Modal
        title='Trip locations'
        visible={modalMap.modalVisible}
        maskClosable={false}
        bodyStyle={{height: '500px'}}
        width='500px'
        onCancel={handleModalOk}
        footer={[
          <Button
              key="back"
              type="primary"
              onClick={handleModalOk} >
              Close
          </Button>
        ]} >
        <MyMapContainer
            searchMode={false}
            tripLocations={modalMap.tripLocations}
            mapCenterLat={mapCenterLat}
            mapCenterLng={mapCenterLng} />
      </Modal>
      </div>
      )
}
export default DisplayTrips