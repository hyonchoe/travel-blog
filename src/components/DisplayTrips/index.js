import React, { useState, useRef } from 'react'
import { useAuth0 } from '@auth0/auth0-react'
import { Empty, Skeleton, Row, Col, BackTop, message, Modal, Button, List, Typography } from 'antd'
import MyMapContainer from '../MyMapContainer'
import Trip from '../Trip'
import useTripData from './helpers/useTripData'

const DisplayTrips = (props) => {    
    const { tripList, reloadTripData, handleDeleteTrip, onLoadMore } = useTripData(props.showMyTrips)
    const { isAuthenticated, user } = useAuth0()
    const [modalMap, setModalMap] = useState({
      modalVisible: false,
      tripLocations: null,
    })
    const [scrollToTripId, setScrollToTripId] = useState('')
    const lastItemRef = useRef(null)

    const onLoadMoreClicked = async () => {
      const { noMoreRecords, scrollToTripId } = await onLoadMore()

      if (noMoreRecords){
        message.info('No more older public trips to display')
      }
      if (scrollToTripId !== -1){
        setScrollToTripId(scrollToTripId)
        if (lastItemRef.current){
          lastItemRef.current.scrollIntoView()
        }
      }
    }
    const handleLaunchMapModal = (tripTitle, tripLocations) => {
      setModalMap({
        modalVisible: true,
        tripLocations: tripLocations
      })
    }
    const handleModalOk = () => {
      setModalMap({
        modalVisible: false,
        tripLocations: null
      })
    }

    const tripCountsPerPage = 10
    const showMyTrips = props.showMyTrips
    const handleEditTrip = props.editTrip
    let mapCenterLat = null
    let mapCenterLng = null
    if (modalMap.tripLocations && modalMap.tripLocations.length>0) {
      mapCenterLat = modalMap.tripLocations[0].latLng[0]
      mapCenterLng = modalMap.tripLocations[0].latLng[1]
    }
    const userId = (isAuthenticated) ? user.sub : ''
    const msgColLayoutSides = {
      xs: { span: 0 },
      md: { span: 2 },
      lg: { span: 3 },
      xl: { span: 4 },
      xxl: { span: 5 }
    }
    const msgColLayoutContent = {
      xs: { span: 24 },
      md: { span: 20 },
      lg: { span: 18 },
      xl: { span: 16 },
      xxl: { span: 14 }
    }
    const tripListColLayoutSides = {
      xs: { span: 0 },
      md: { span: 3 },
      lg: { span: 4 },
      xl: { span: 5 },
      xxl: { span: 6 }
    }
    const tripListColLayoutContent = {
      xs: { span: 24 },
      md: { span: 18 },
      lg: { span: 16 },
      xl: { span: 14 },
      xxl: { span: 12 }
    }

    const loadMoreButton = (!tripList.loadingData && !tripList.noMoreRecords) ? 
                              (<div style={{ textAlign: 'center' }} >
                                <Button onClick={onLoadMoreClicked}>Load more</Button>
                              </div>)
                              : null
    const tripCard = (item, itemRef) =>
                              (<div ref={itemRef} className="tripContainer">
                                { item.placeholder && 
                                <div style={{ textAlign: 'center' }} >
                                  <Button onClick={reloadTripData}>Trips removed from display. Click to refresh the feed</Button>
                                </div>
                                }
                                { !item.placeholder &&
                                <Trip
                                  isAuthenticated={isAuthenticated}
                                  userId={userId}
                                  showMyTrips={showMyTrips}
                                  trip={item}
                                  deleteTrip={handleDeleteTrip}
                                  editTrip={handleEditTrip}
                                  launchMapModal={handleLaunchMapModal} />
                                }
                              </div>)

    return (
      <div className="displayTripsContainer">
        <BackTop />
          <Row
            gutter={[0, 8]}
            justify="start" >
            <Col {...msgColLayoutSides} />
            <Col {...msgColLayoutContent} >
              <Typography.Title>{greetingMsg(showMyTrips)}</Typography.Title>
            </Col>
            <Col {...msgColLayoutSides} />
          </Row>

          <Row
            gutter={[0, 8]}
            justify="center" >
            <Col {...tripListColLayoutSides} />
            <Col {...tripListColLayoutContent} >
              { !showMyTrips && tripList.trips.length > 0 && 
              <List
                itemLayout="vertical"
                loading={tripList.loadingData}
                loadMore={loadMoreButton}
                dataSource={tripList.trips}
                renderItem={(item) => (
                  <List.Item>
                    {item._id === scrollToTripId && 
                      tripCard(item, lastItemRef)
                    }
                    {item._id !== scrollToTripId && 
                      tripCard(item, null)
                    }
                  </List.Item>
                )} />
              }
              { showMyTrips && tripList.trips.length > 0 && 
              <List
                itemLayout="vertical"
                pagination={{
                  onChange: page=> {
                    window.scrollTo({top:0, behavior: 'smooth'})
                  },
                  pageSize: tripCountsPerPage,
                }}
                dataSource={tripList.trips}
                renderItem={(item) => (
                  <List.Item>
                    {tripCard(item)}
                  </List.Item>
                )} />
              }
              { tripList.loadingData &&
              <Skeleton loading={true} active />
              }
              { tripList.trips.length === 0 && !tripList.loadingData &&
              <Empty />
              }
            </Col>
            <Col {...tripListColLayoutSides} />
          </Row>
        <Modal
          title='Trip locations'
          visible={modalMap.modalVisible}
          maskClosable={false}
          bodyStyle={{height: '500px'}}
          width='500px'
          onCancel={handleModalOk}
          footer={[
            <Button
                key="back"
                type="primary"
                onClick={handleModalOk} >
                Close
            </Button>
          ]} >
          <MyMapContainer
              searchMode={false}
              tripLocations={modalMap.tripLocations}
              mapCenterLat={mapCenterLat}
              mapCenterLng={mapCenterLng} />
        </Modal>
      </div>
      )
}

const greetingMsg = (isMyTrips) => {
  if (isMyTrips){
    return 'Your memorable travels and memories'
  }
  return "Stories and footmarks from travelers in the world"
}

export default DisplayTrips