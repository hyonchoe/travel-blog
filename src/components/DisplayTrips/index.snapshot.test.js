/**
 * Snapshot tests for DisplayTrips component.
 * 
 * Run by running 'npm test' in command line.
 */

import React from 'react'
import { shallow } from 'enzyme'
import DisplayTrips from './'

describe('DisplayTrips', () => {
    const handleEditTrip = jest.fn()
    
    it('matches snapshot for public trips', () => {
        const wrapper = shallow(
            <DisplayTrips
                showMyTrips={false}
                editTrip={handleEditTrip} />
        )
        expect(wrapper).toMatchSnapshot()
    })

    it('matches snapshot for my trips', () => {
        const wrapper = shallow(
            <DisplayTrips
                showMyTrips={true}
                editTrip={handleEditTrip} />
        )
        expect(wrapper).toMatchSnapshot()
    })
})