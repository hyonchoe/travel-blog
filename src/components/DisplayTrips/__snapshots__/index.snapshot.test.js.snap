// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayTrips matches snapshot for my trips 1`] = `
<div
  className="displayTripsContainer"
>
  <Memo(BackTop) />
  <Row
    gutter={
      Array [
        0,
        8,
      ]
    }
    justify="start"
  >
    <Col
      span={2}
    />
    <Col
      span={20}
    >
      <Title>
        Your memorable travels and memories
      </Title>
    </Col>
    <Col
      span={2}
    />
  </Row>
  <Row
    gutter={
      Array [
        8,
        16,
      ]
    }
    justify="center"
  >
    <Col
      span={4}
    />
    <Col
      span={16}
    >
      <Skeleton
        active={true}
        avatar={false}
        loading={true}
        paragraph={true}
        title={true}
      />
    </Col>
    <Col
      span={4}
    />
  </Row>
  <Modal
    bodyStyle={
      Object {
        "height": "500px",
      }
    }
    confirmLoading={false}
    footer={
      Array [
        <ForwardRef(InternalButton)
          block={false}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
          type="primary"
        >
          Close
        </ForwardRef(InternalButton)>,
      ]
    }
    maskClosable={false}
    maskTransitionName="fade"
    okType="primary"
    onCancel={[Function]}
    title="Trip locations"
    transitionName="zoom"
    visible={false}
    width="500px"
  >
    <MyMapContainer
      mapCenterLat={null}
      mapCenterLng={null}
      searchMode={false}
      tripLocations={null}
    />
  </Modal>
</div>
`;

exports[`DisplayTrips matches snapshot for public trips 1`] = `
<div
  className="displayTripsContainer"
>
  <Memo(BackTop) />
  <Row
    gutter={
      Array [
        0,
        8,
      ]
    }
    justify="start"
  >
    <Col
      span={2}
    />
    <Col
      span={20}
    >
      <Title>
        Stories and footmarks from travelers in the world
      </Title>
    </Col>
    <Col
      span={2}
    />
  </Row>
  <Row
    gutter={
      Array [
        8,
        16,
      ]
    }
    justify="center"
  >
    <Col
      span={4}
    />
    <Col
      span={16}
    >
      <Skeleton
        active={true}
        avatar={false}
        loading={true}
        paragraph={true}
        title={true}
      />
    </Col>
    <Col
      span={4}
    />
  </Row>
  <Modal
    bodyStyle={
      Object {
        "height": "500px",
      }
    }
    confirmLoading={false}
    footer={
      Array [
        <ForwardRef(InternalButton)
          block={false}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
          type="primary"
        >
          Close
        </ForwardRef(InternalButton)>,
      ]
    }
    maskClosable={false}
    maskTransitionName="fade"
    okType="primary"
    onCancel={[Function]}
    title="Trip locations"
    transitionName="zoom"
    visible={false}
    width="500px"
  >
    <MyMapContainer
      mapCenterLat={null}
      mapCenterLng={null}
      searchMode={false}
      tripLocations={null}
    />
  </Modal>
</div>
`;
